src/Marten/Events/Daemon/Internals/SubscriptionAgent.cs

```diff
@@ -0,0 +1,302 @@
+    public ShardName Name { get; }
+    private readonly CancellationTokenSource _cancellation = new();
+    private readonly ActionBlock<Command> _commandBlock;
+    private ErrorHandlingOptions _errorOptions = new();
+    private IDaemonRuntime _runtime = new NulloDaemonRuntime();
+
+    public SubscriptionAgent(ShardName name, AsyncOptions options, IEventLoader loader,
+        ISubscriptionExecution execution, ShardStateTracker tracker, ILogger logger)
+    {
+        _options = options;
+        _loader = loader;
+        _execution = execution;
+        _tracker = tracker;
+        _logger = logger;
+        Name = name;
+
+        _commandBlock = new ActionBlock<Command>(Apply, _cancellation.Token.SequentialOptions());
+
+        ProjectionShardIdentity = name.Identity;
+        if (_execution.DatabaseName != "Marten")
+        {
+            ProjectionShardIdentity += $"@{_execution.DatabaseName}";
+        }
+    }
+
+    public string ProjectionShardIdentity { get; private set; }
+
+    public CancellationToken CancellationToken => _cancellation.Token;
+
+    // Making the setter internal so the test harness can override it
+    // It's naughty, will make some people get very upset, and
+    // makes unit testing much simpler. I'm not ashamed
+    public long LastEnqueued { get; internal set; }
+
+    public long LastCommitted { get; internal set; }
+
+    public long HighWaterMark { get; internal set; }
+
+    public async Task ReportCriticalFailureAsync(Exception ex)
+    {
+        try
+        {
+#if NET8_0_OR_GREATER
+            await _cancellation.CancelAsync().ConfigureAwait(false);
+#else
+            _cancellation.Cancel();
+#endif
+            await _execution.HardStopAsync().ConfigureAwait(false);
+            PausedTime = DateTimeOffset.UtcNow;
+            Status = AgentStatus.Paused;
+            _tracker.Publish(new ShardState(Name, LastCommitted) { Action = ShardAction.Paused, Exception = ex});
+
+            if (Mode == ShardExecutionMode.Rebuild)
+            {
+                _rebuild?.SetException(ex);
+            }
+
+            await DisposeAsync().ConfigureAwait(false);
+        }
+        catch (Exception e)
+        {
+            _logger.LogError(e, "Error trying to pause subscription agent {Name}", ProjectionShardIdentity);
+        }
+    }
+
+    long ISubscriptionAgent.Position => LastCommitted;
+
+    public AgentStatus Status { get; private set; } = AgentStatus.Running;
+
+    public async Task StopAndDrainAsync(CancellationToken token)
+    {
+        try
+        {
+            // Let the command block finish first
+            _commandBlock.Complete();
+            await _commandBlock.Completion.ConfigureAwait(false);
+
+#if NET8_0_OR_GREATER
+            await _cancellation.CancelAsync().ConfigureAwait(false);
+#else
+            _cancellation.Cancel();
+#endif
+
+            await _execution.StopAndDrainAsync(token).ConfigureAwait(false);
+        }
+        catch (TaskCanceledException)
```

Comment:
```suggestion
        catch (OperationCanceledException)
```
Because of https://github.com/dotnet/runtime/issues/83406